### ISSUE TITLE
Support for autotile (RPG Maker format)

### ISSUE THREAD

**Issue by Frozenfire92** [bjorn/tiled#1022]
Tiled is a great tool, especially for those who use Unity or game engines other than RPG maker. With the release of the newest humble bundle I foresee many people like myself wanting to use their new assets with Tiled (and eventually game engines other than RPG maker). It appears that their `Autotile` format is quite common and works well for compacting things, but makes it difficult to use in Tiled without an external tool.

I think it is within the scope of Tiled to support importing this format and then automatically slicing it up. While the terrain tool works well for some sprites, it doesn't seem to work with this format. There is an existing tool [Remex](https://www.assembla.com/code/rpg-maker-to-tiled-suite/subversion/nodes) but it only supports the 64x96 pixel image at a time. It would be better to automatically detect the size and split it accordingly (It appears 512x384 is standard)

**Comment by bjorn** [comment]
It would definitely be interesting to support this type of terrain, though I have not seen it used outside of RPG Maker so I kind of disagree with the format being "quite common". Outside of RPG Maker, the format used by Tiled is actually quite common. In any case the question comes up a lot, I guess mainly because of the vast library of tilesets available alongside (or made for use with) RPG Maker.

I'm not sure what's the best way to support it. There are some options:

* Ship a tool like Remex with Tiled. However, it is plain annoying to have to preprocess your tilesets and it needlessly increases the texture size.

* Build into Tiled some intelligence about how to cut and recombine these autotiles. Unfortunately, when we go this way then this kind of code would need to be added to all libraries and frameworks out there that support Tiled maps.

* Half the tile size so that Tiled maps can stay simple and compatible with existing libraries and frameworks, but allow the tools to apply RPG Maker stile automapping. This increases the size of the layer data by four and puts the complexity of dealing with the halved tile size in the UI. I guess this last approach is the one used by RPG Maker itself, but I do not know its map format.

In any case, if you want to help me get around to a feature like this, please do consider [supporting me on Patreon](https://www.patreon.com/bjorn). The amount of time I have for Tiled directly depends on how many hours I can afford to drop from my job.

**Comment by Frozenfire92** [comment]
Thanks for the quick response,

Forgive me in being a bit uneducated in the commonality of this format. I think a lot of people like myself have bought the humble bundle and now have a ton of assets in a format for an engine they aren't interested in using. 

> Ship a tool like Remex with Tiled. However, it is plain annoying to have to preprocess your tilesets and it needlessly increases the texture size.

I think this is the best option. When creating a new tileset the dropdown for type could be `Based on RPG Maker tileset` which would then auto-detect and split into 64x96 sized chunks to be used by Remex (or the forked code integrated). I think this would make the process much smoother than it is now (cutting the image up and running remex on each individually before loading into tiled)

I thank you for your time you have invested in this project, it is very useful :smile:  I will try and have a peek a the code for it and Remex over the next week and look into helping with this feature.

**Comment by lefticus** [comment]
I bought the Game Maker Humble Bundle myself and feel compelled to point out that much (most?) of the material in it has a Copyright statement such as:

> The graphics included in the Product (the "Pirate Ship Tile Pack") may be used with any of ENTERBRAIN's game development software (include all related titles of "RPG MAKER Series" or "our products") that retains the Company's rights. You cannot use the Resource Pack with other game engines/software. User can create, distribute, or sell their original game by using the Resource Pack only if they meet all the following conditions:  

If I read that correctly that means we cannot legally use these tilesets with tools like "tiled" anyhow.

This is now the second Humble Bundle with RPG Maker in it that I've bought and I've seen many copyright licenses similar to this one.

**Comment by bjorn** [comment]
Right, that's the deal with RPG Maker's own tilesets unfortunately, so supporting them directly is pretty much only useful for 3rd party tilesets that were painted to be RPG Maker compatible.

**Comment by Frozenfire92** [comment]
I believe they addressed that here https://support.humblebundle.com/hc/en-us/articles/206156508 . I think limiting assets to a specific engine goes against what humble bundle is about anyhow.

**Comment by lefticus** [comment]
@Frozenfire92 Ah, thank you. I was pretty confused by that.

**Comment by bjorn** [comment]
@Frozenfire92 That's great to know!

**Comment by fmoo** [comment]
Just wanted to add a link to some docs on [how autotiles work in RMVXA](http://blog.rpgmakerweb.com/tutorials/anatomy-of-an-autotile/).

Like `Remex` (which is basically abandoned these days), that link only covers the most popular "A2" format, so I'll to provide some extra details from my investigation on the other "A" formats.

- The "A1" format has the same number of tiles as A2, except the first three (column-wise) blocks out of every four blocks of microtiles are animations of each other (e.g., water tiles). 
- The "A3" format uses squre blocks of 16 minitiles for every block instead of the (rectangular) 24 minitiles from A1/A2 and as a result basically supports rendering of square/rectangular surfaces.  This is used primarily for things like walls, steeped roofs, bridges, etc.
- The "A4" format is a hybrid of A2 and A3, where below every block of 24-minitile surfaces, there is a 16-minitile square block for walls themed according to the surface.
- "A5" appears to be a generic, non-minitile based tilesheet.
- "B" and "C" are for objects, and are generic, non-minitile based tilesheets as well.

**Comment by kphillisjr** [comment]
I also think this is a good idea, and will greatly expand on what @fmoo stated. Also I will say that the following information can be used by anyone wishing to import tiles into this map editor.

## Tile Types

### Generic Terrain autotile
These are the easiest to work with, and cover most of the auto-tiles. For RPGMaker VX Ace optimized tilesets these are a 16 pixel by 16 pixel grid. I know my example is smaller than this. The Blue area represents the edge case with other tiles, and the white area is the terrain type for the Auto-Tile. 
![coreautotile](https://cloud.githubusercontent.com/assets/1205559/21493858/5b5ce5e4-cbd3-11e6-94cb-cbc3448cdc59.png)

### Animated Terrain autotile
These are terrain auto-Tiles from above where you have 3 frames of animation side by side.

### Generic Wall Auto-Tile
This works exactly the same as Terrain auto-Tiles, but lacks the small inner corner case. In General these tiles do not have interaction with other tile-types.
![wallautotile](https://cloud.githubusercontent.com/assets/1205559/21494048/59610760-cbd4-11e6-9e20-abdb8217710e.png)

### WaterFall Auto-Tiles
This tile type is a set of two 32px by 32px animated tiles side by side where each row is a frame of animation. This means that this tile takes up the same amount of space on a sprite sheet as a generic terrain auto-tile. This can form an animated wall auto-tile when you have 3 or more that can be linked together.  In general this is the type of tile where the users need to configure on a set-by-set basis because this can involve 3 types of terrain ( Waterfall, Water, and land type )

## RPG Maker Tile Layers
### Tile Layer A1
This is the most complex of the sprite sheets, but in general all of these have the similar layout as shown below:
![autotilea](https://cloud.githubusercontent.com/assets/1205559/21494978/9b9b0e90-cbda-11e6-85a6-0957f39b30fb.png)
#### Tile Layer A1 - Area A - Water/Ocean
This is a Animated Terrain Auto-tile. In general it is used for water.
#### Tile Layer A1 - Area B - Deep Water/Ocean
This is a Animated Terrain Auto-tile. In general it is used for deeper water, and sometimes is placed within Area A.
#### Tile Layer A1 - Area C - More animated Water Tiles
More Animated Terrain Auto-Tiles. To my knowledge these have been used for Lava, Ice, and even more water animations.
#### Tile Layer A1 - Area D - Water Tile decoration
These are generally used for decoration of water tiles in an auto-terrain format.
#### Tile Layer A1 - Area E - More Water Decoration Tiles
These could be terrain auto-tiles or waterfall tiles. Either way this section requires a lot of user input, but in general these are always two 32px by 32px tiles side by side, and sometimes 3 of these can be combined to form a wall auto-tile.

### Tile Layer A2
This is a straight forward import, this is a grid of 8 by 4 terrain auto-tiles.

### Tile Layer A3 - Exterior wall/Ceiling
This is used for exterior Ceiling/walls, and is a 8 by 4 grid of wall auto-tiles. I believe the easiest solution for now is to define each wall auto-tile  as a terrain.

### Tile Layer A4 - Wall/Ceiling
This is is simply a row of 8 terrain auto-tiles used for Ceilings followed by a row of 8 wall auto-tiles. In general this forms a total of 48 tiles total.

### Tile Layer A5 - Generic Tiles
These are 32px by 32px tiles. 

### Tile Layer B to E - Generic Tiles
These are Objects of varying size.


## Possible issues.
### Animated Terrains
I have not found any simple way to create animated terrains yet, and this is a required feature for this bug report.

### Cross-File Terrains
I have found that occasionally the terrain auto-tiles have the center fill-tile found in the TileA5 file. To my knowledge there is no easy way to resolve this issue without editing the input files directly.

### Terrain Fill
There is no easy way to fill using defined terrains.  Instead you need to fill using tiles in the tile sheet directly.

## Possible Solution
There is many solutions for this, but the easiest would be to use import wizard. The Wizard will need to have the user help guide the import since there is going to be a large set of tiles imported in one go. Most of the work would be terrain definition/names, and the occasional cross-file terrain fix.

**Comment by no-identd** [comment]
Some semi-recent discussion of this issue on itch:

https://itch.io/post/487538

**Comment by no-identd** [comment]
#1421 has now gotten closed as #2708 by @phlosioneer has landed. This still has a catch tho, to quote @bjorn:
>I think the main catch right now is that automatic updating of the meta-tileset image will only be triggered when its map file is changed. Changes to any images or other meta-tilesets referenced by the map file will not trigger such an update.

(I don't think a follow up issue exists for this yet, but I assume one will soon arise.)

However, either way, how much closer does that bring this issue to finding a resolution? 🤔

**Comment by bjorn** [comment]
@no-identd Well, since RPG Maker performs its autotile magic by cutting the tiles into even smaller pieces and recombining them, this could now be done using a map and then that map can be used as the image of a tileset, where the terrain information for the terrain tool can then be set up.

Of course, that's a rather manual process that could also have been done using any image editing software. But a big time-saver is that the map of this tileset as well as the terrain information can be re-used with different RPG Maker tilesets, so it only needs to be set up once.

**Comment by bjorn** [comment]
Just a note to those wanting to use RPG Maker autotiles in Tiled: @fmoo wrote a tool for converting such tiles to a format that Tiled's Terrain tool can work with, available at https://fmoo.itch.io/autotile-packing-tools.

**Comment by devium** [comment]
I wasn't happy with the tools available, so I wrote another conversion tool that pre-combines RPG Maker's smaller subtiles into all possible combinations:
https://github.com/devium/tiled-autotile
It's so far just a quick and dirty implementation, lacks documentation, config options, and only works with tileset layers in the specific A1 to A4 formats as explained by @kphillisjr above. But it gets the job done for now.

Basically, it turns this:
![image](https://user-images.githubusercontent.com/6115773/111684680-4e147e00-8827-11eb-9c43-fb1d91c67b7e.png)
into this:
![image](https://user-images.githubusercontent.com/6115773/111684716-5a004000-8827-11eb-9418-a06d6a60d37b.png)
So it also makes single-tile "rivers" and paths available and arranges them a little neater.

**Comment by bjorn** [comment]
@devium That looks great! Since you're creating blobs, I wanted to try it out with the recently improved Terrain Brush, and it worked great:

![image](https://user-images.githubusercontent.com/531764/111692049-478a0480-882f-11eb-829f-52b5d825247a.png)

Eventually it would still be nice if we could avoid the explicit expanding, though actually we would not get around to that without either implementing the sub-tile logic in Tiled as well as in the various game engines, or saving to a tile layer with smaller tiles (which depends on supporting tile layers with differently sized tiles in general).

Also unfortunately the single-tile is currently not covered by the Terrain Brush. Hopefully we can find a solution to this some time.

**Comment by eishiya** [comment]
I made a version of @devium's script that runs directly in Tiled (requires 1.8 RC or newer): https://github.com/eishiya/tiled-expand-autotile
This one outputs a Tiled map, which can be exported as an image, or saved to use as a Tileset source directly. The output layout is the same as in the original script.

I'm hoping that with a script like this, Metatilesets, and another script to import terrains from the original tileset(s) into the metatileset, and hopefully importer/exporter support for metatilesets, using RPG Maker tilesets in Tiled can become a lot less painful. Unfortunately, I don't foresee wide support for metatilesets. However, they can be pretty painlessly replaced with an image export of the source map for production use. ...and engine support for metatilesets seems more likely than engine support for RPG Maker-style minitiles outside of RPG Maker :/

**Comment by bjorn** [comment]
@eishiya That's great! Now if only we could also find a way to automatically set up the terrains. Problem of course is that then you need to create the tileset, for which you'd first need to save the image. Could be added as a separate action I guess.

**Comment by eishiya** [comment]
In the README I mentioned wanting to write a script that imports and expands labels from the original tileset, since that's a lot less work for the user than labelling the expanded tileset, but still gives the user control over what's what. It would be a separate script, since it's useful beyond just RPG Maker tilesets. It would only work:
- on Metatilesets, so that there's the destination tileset, the map, and the original tileset all available. If the user wants to use an image, they'd need to make a metatileset first, run the action, and then replace the source map with the image.
- when the metatile size is exactly twice the subtile size. I suppose I could make it work for other cases as long as the metatile edges align with subtile edges, but I wonder if anyone even makes metatilesets like that?

Do you think that's worth taking the time to write?

It would've been cool to be able to generate the terrains directly as part of the autotile expanding script, but as you said, it would require saving the image/map first to create the tileset. I could write the terrain import script above, and then write a bridge script that the user can run after saving the map that generates the tileset, runs the terrain import action, and outputs the tileset, I suppose...

**Comment by CelticMinstrel** [comment]
Having fiddled with eishiya's script and a bunch of other things related to this, I am now convinced that the best way to handle this is, essentially, the way eishiya's script does it… but natively supported in the format.

Essentially: when creating the tileset, allow choosing the _base image_ and the _template map_. This template map is literally what eishiya's script outputs, mapping the subtiles of the base image to their proper locations in the full blob or edge pattern. Simply substituting the chosen base image into that map (and also setting the tile sizes to match) will produce a working tileset. The tilesetfile would then have separate references to the two files – they'd be composed on the fly as needed.

I [created a script](https://discourse.mapeditor.org/t/support-for-rpgmaker-style-compressed-tile-layouts/7796/1) to demonstrate what I think it could look like. This proves that it can be done by scripting, so you could argue that, perhaps, it's better to just leave it to a script. But I think adding it to the tileset creation window would be a better flow than needing to use an entirely different "New Tileset" command whenever you want to use subtile patterns.

The nice thing about this approach is that it easily generalizes to different types of patterns. You can have templates for A1, A2, A3, and A4 format tilesheets. And you can have templates for XP and 2k(3) tilesheets, which are very different in form. And you can have templates for custom layouts like "microblob". You could even have templates for systems that allow the subtiles to be rotated or reflected, or exotic systems that use more than 4 subtiles, if you wanted. Maybe you could even support template tilesets that have multiple slots to substitute tileset images into. And nothing needs to change on the gamedev end – at most, the export process might need to render that metatileset to an image because the chosen framework doesn't support metatilesets (which, unfortunately, it seems most don't currently).

It's even possible to encode terrain and animation information into the template, by placing it in the embedded tileset – that data won't make it directly into the resulting tileset, but it can be used to construct the proper terrains and animations. My script already does that, though it's probably somewhat inefficient at it and may not be correct in some edge cases. You could also do this by making the template be a tileset instead of a tilemap, which would eliminate the need to synthesize the terrains and animations (they can just be taken as-is), but that would become a little unwieldy unless you could directly embed the map data into the tileset file.

And the most common subtile formats (namely, those used by various versions of RPGMaker) could easily be shipped with Tiled if desired. If not, they could be distributed as an easily-downloadable pack somewhere. I already made templates for all RPGMaker formats I know of.

Maybe templates could even be configured to support additional options. This seems like it might be getting too complicated, but I've found two cases that don't quite fit nicely in the above framework. The first is that, unlike every other RPGMaker, Vx(Ace) apparently supports autotile animations of arbitrary length, rather than always 4 frames. So, if you have an autotile that happens to have animations of a different length, you'll need to edit the template to add or remove frames. The second is that RPGMaker 2k(3) apparently let you choose between 3-frame or 4-frame animation for the water autotiles. For this case could just have separate 3-frame and 4-frame templates.
